### DISCUSSION TITLE
Any idea to publish the rome-rowan to crates.io

### DISCUSSION THREAD

**Discussion by IWANABETHATGUY** [rome/tools#1778] [Suggestions]
Recently i am writing a transpiler like postcss, i try to use rowan, but mutate the syntax node is pretty annoying, i found that your goal is samily like me, so I am wonder if you have any plan to publish the your `rowan` , so other could use it for transpiler like stuff

**Comment by ematipico** [comment]
Yeah it might be something that we could do. Although we are still heavily working on it: adding `Trivia`s, slots (recently merged), mutation APIs (which actually creates new trees). @jamiebuilds what do you think?

**Reply by IWANABETHATGUY** [reply]
recently I found rowan's mutation api is quite inefficient, every mutation, they would recreate a new tree, when CST grow, the performance would be really poor, any thought about this?

**Reply by IWANABETHATGUY** [reply]
maybe not the whole tree, I mean `clone_for_update` would recrate the partial tree.

**Reply by ematipico** [reply]
Yeah this is something we are exploring at the moment. @yassere crated a draft PR to explore it: https://github.com/rome/tools/pull/1773

**Reply by yassere** [reply]
We haven’t ruled out the possibility of using rowan’s mutation API, but it’s true that we’re leaning towards sticking with immutable trees.

I don’t think we’ve made many decisions yet about which of our crates we plan to maintain as independent libraries. There’s going to be lots of code churn in the near future, so I assume we’ll put more thought into that once things start to stabilize.

**Reply by IWANABETHATGUY** [reply]
> We haven’t ruled out the possibility of using rowan’s mutation API, but it’s true that we’re leaning towards sticking with immutable trees.
> 
> I don’t think we’ve made many decisions yet about which of our crates we plan to maintain as independent libraries. There’s going to be lots of code churn in the near future, so I assume we’ll put more thought into that once things start to stabilize.

I'am curious if you don't use the mutable api how do you tranform the ast, emit on the fly during transforming?

**Reply by yassere** [reply]
We're exploring different approaches for transforming the syntax tree and it's possible we'll end up with more than one depending on the use case. The draft PR #1773 (probably on hold until next week) is one possible approach that allows for making several non-overlapping edits to a tree and rebuilding it only once. It's likely that would still need to be paired with something like the Make/With/Builder APIs from [Swift libSyntax](https://github.com/apple/swift/tree/main/lib/Syntax).

**Comment by jamiebuilds** [comment]
I'd like for our rowan crate to get a little bit more mature. Once we start building on top of it, we'll likely ship a number of additional performance/memory/usability improvements. At that point I'd like to talk more with the maintainers of the other rowan crates to see if we can unify anything. At that point we can discuss publishing it
